### ISSUE TITLE
pango 1.50.0 breaks spacing

### ISSUE THREAD

**Issue by marzzzello** [greshake/i3status-rust#1367]
OS: Arch Linux

pango 1.48.10-1:
![pango 1 48 10-1](https://user-images.githubusercontent.com/15198431/144905695-a01efd82-9273-482e-be61-991830c3c646.png)

pango 1.50.0-2
![pango 1 50 0-2](https://user-images.githubusercontent.com/15198431/144905687-3bce8f1c-e41d-4c36-a8f5-e6e5e78883a0.png)

Changelog: https://gitlab.gnome.org/GNOME/pango/-/blob/main/NEWS

**Comment by IndependentFool** [comment]
I would like to add to this as well. Same issue. Nonetheless, downgrading to pango 1.48 fixed it for me.

OS: Arch
i3status-rust: Latest version 0.20.7-1

pango 1.48.10
![pango-1-48-10](https://user-images.githubusercontent.com/59860340/144957396-8f734c8c-4775-4d62-b023-ace350694102.png)

pango 1.50.0
![pango-1-50](https://user-images.githubusercontent.com/59860340/144957422-2367a4da-b039-4578-b1d5-f3a5cff95e7a.png)

**Comment by cfsmp3** [comment]
On first inspection it seems more likely to be an issue with pango (or maybe, i3) than with i3status-rs - the only thing it does about pango is escape some characters. 

BTW some other totally unrelated project is complaining

https://bugs.archlinux.org/task/72939

**Comment by ammgws** [comment]
I guess this is the fix: https://gitlab.gnome.org/GNOME/pango/-/commit/8b46f5db285c6c661a2a7c28c5b354426a8d5c61 ?

**Comment by marzzzello** [comment]
> I guess this is the fix: https://gitlab.gnome.org/GNOME/pango/-/commit/8b46f5db285c6c661a2a7c28c5b354426a8d5c61

pango 1.50.0-3 which includes these commits doesn't fix it for me.

**Comment by epixinvites** [comment]
![image](https://user-images.githubusercontent.com/54717060/145175035-72178d75-e5e7-49c1-ab95-721e88bd6333.png)
`i3status-rs 0.20.8 (commit 64fcdeb 2021-12-03)` along with `pango 1:1.50.0-3`

**Comment by mykeb** [comment]
otf-font-awesome works for me but not ttf-font-awesome. The otf is a bit larger in size though.

**Comment by ammgws** [comment]
Is this happening on both sway and i3? I haven't looked too deeply but was surprised not to find other status bars/etc also complaining about this. All we do is generate JSON which is parsed by swaybar/i3bar which are what actually renders the text.

Perhaps try running a small program like this against the "good" and "bad" versions of pango and see what results you get?

Or otherwise run a git bisect on pango while testing the bar each time to find the commit that changed the behaviour.

```c
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <stdio.h>
#include <cairo.h>
#include <pango/pangocairo.h>

int main (int argc, char **argv)
{
  if (argc != 3)
  {
     return 1;
  }
  char *text = argv[1];
  char *font = argv[2];
  cairo_surface_t *target = cairo_recording_surface_create(CAIRO_CONTENT_COLOR_ALPHA, NULL);
  cairo_t *cairo = cairo_create (target);
  PangoLayout *layout = pango_cairo_create_layout(cairo);
  PangoFontDescription *desc = pango_font_description_from_string(font);
  pango_layout_set_font_description(layout, desc);
  pango_layout_set_text(layout, text, -1);
  int text_width, text_height;
  pango_cairo_update_layout(cairo, layout);
  pango_layout_get_pixel_size(layout, &text_width, &text_height);
  char *pango_font_used = pango_font_description_to_string(desc);
  printf("Width of '%s' is %d\nUsing font: %s", text, text_width, pango_font_used);
}
```

**Comment by ammgws** [comment]
You can use `pango-view` to test this as well, try playing around with it: `pango-view --text "stuff here" --font FontAwesome`

**Comment by codicodi** [comment]
I've reported this upstream [1]. If someone has some more insight that might help pango developers understand the issue, please do share.

[1] https://gitlab.gnome.org/GNOME/pango/-/issues/641

**Comment by ammgws** [comment]
Thanks for reporting it. Looks like it should be fixed now: https://gitlab.gnome.org/GNOME/pango/-/commit/37f22bb80ebacd834f1927e6acd7fd94aeaf1270

**Comment by c02y** [comment]
It seems more spacing is added comparing to the original version.
1.48.9: one space
![1_48_9](https://user-images.githubusercontent.com/2135996/146619799-50814277-59fc-477e-a940-92fb093ae8a6.png)

1.50.1: no space
![1_50_1](https://user-images.githubusercontent.com/2135996/146619817-c974c1b6-d890-464e-8f68-8794e8876488.png)

1.50.2: two spaces
![1_50_2](https://user-images.githubusercontent.com/2135996/146619823-d03d4b0e-2e43-4b19-ba1b-da1f619e275a.png)

**Comment by matthiasclasen** [comment]
pango perspective here:

If you want control about the amount of space, use a font that has a space glyph.

All of these icon fonts are weird and deficient in various ways, and forcing a text layout system like pango to deal with them is only ever going to yield suboptimal results. Pango makes various assumptions about fonts which are quite reasonable for
text fonts, such as: the font shall have a space glyph, or at least it shall have hex digits so we can devise some reasonable
width for space by measuring those. These icon fonts I've been dealing with in the last few days don't.

**Comment by c02y** [comment]
> @ammgws It seems more spacing is added comparing to the original version. v1.48.9: 1 space 
> ![image](https://user-images.githubusercontent.com/2135996/146620236-09003014-dbb6-45cc-888c-260ffeffaeaf.png)
> 
> v1.50.1: no space 
> ![image](https://user-images.githubusercontent.com/2135996/146620275-45892169-07b7-4f00-8d24-75039e1fb1e8.png)
> 
> v1.50.2: 1.5 or 2 spaces ![image](https://user-images.githubusercontent.com/2135996/146620306-23817102-31da-4855-95cc-1e8f6bfcb584.png)

Just update new version v1.50.3 under Archlinux, it seems good to me.
![image](https://user-images.githubusercontent.com/2135996/147097739-0dee2b8e-fffe-485c-991f-81fdc5131329.png)

**Comment by ammgws** [comment]
Thanks for the update, looks like it was this: https://gitlab.gnome.org/GNOME/pango/-/commit/9fe2853692bb64ded58a507d87508dc8536cd386 that fixed it. And thanks to Matthias for the persistent effort!

**Comment by marzzzello** [comment]
yes, 1.50.3 fixed it
